### PR TITLE
FIX: Special chars in HID names causing rebinding to not work (case 1335465).

### PR DESCRIPTION
**Pull request by Rene-Damm**
Fixes [1335465](https://issuetracker.unity3d.com/issues/general-certain-hid-joysticks-fail-to-send-input-on-windows-even-though-interactive-rebind-works) ([FogBugz](https://fogbugz.unity3d.com/f/cases/1335465/)).

### Description

In control paths, characters like (/, <, >, {, and }) take on special meaning. Backslashes are used to escape this.

However, the current code takes many strings (such as product names) as is and puts them into control paths where they are then read incorrectly.

### Changes made

The code now escapes and unescapes strings in several places. Also, the / character takes on special meaning only while not inside a {..} or <..> group.

### Notes

`InputControlPath` is still waiting for a proper refactor which I chose not to roll into this bugfix. There's the `PathParser` code now and `MatchPathComponent` should be brought over to it. All code operating on path structures should be rolled over into using `PathParser`.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControlPath.cs [hunk 1]
```diff
@@ -97,6 +97,19 @@ public static class InputControlPath
 
         public const char Separator = '/';
 
+        // We consider / a reserved character in control names. So, when this character does creep
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControlPath.cs:100]: What if the original string contains space?

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControlPath.cs:100]: Maybe its less likely that this string UTF-8(?) from lower character table is not present or are we sure there is no "ids" containing space?

**Review by ekcoh — APPROVED** [review]
Approved but one question/concern regarding using space as replacement char, maybe we already got it covered?

File: Assets/Tests/InputSystem/Plugins/HIDTests.cs [hunk 2]
```diff
@@ -1312,5 +1311,54 @@ public void Utilities_CanRecognizeVendorDefinedUsages()
...
         Assert.That(HID.UsageToString((HID.UsagePage) 0xff01, 0x33), Is.Null);
     }
+
+    // https://fogbugz.unity3d.com/f/cases/1335465/
+    [Test]
+    [Category("Devices")]
+    [TestCase('/')]
+    [TestCase('\\')]
+    [TestCase('>')]
+    [TestCase('<')]
```
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/HIDTests.cs]: Maybe add a test case containing a space?

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControlExtensions.cs [hunk 3]
```diff
@@ -912,22 +912,24 @@ internal static string BuildPath(this InputControl control, string deviceLayout,
             var device = control.device;
 
             builder.Append('<');
-            builder.Append(deviceLayout);
+            builder.Append(deviceLayout.Replace("\\", "\\\\").Replace(">", "\\>"));
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Controls/InputControlExtensions.cs]: Curious why we don't use our own StringHelper.Escape method for all this escaping?

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControlPath.cs (hunk 1 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControlPath.cs:100]: Spaces are fine. They have no special meaning in control paths.

File: Packages/com.unity.inputsystem/InputSystem/Controls/InputControlExtensions.cs (hunk 3 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Controls/InputControlExtensions.cs]: Fixed

File: Assets/Tests/InputSystem/Plugins/HIDTests.cs (hunk 2 above)
Comment (Rene-Damm) [reply Assets/Tests/InputSystem/Plugins/HIDTests.cs]: Done

**Review by andrew-oc — APPROVED** [review]

**Pull request merged**